motoman_msgs/srv/SelectTool.srv

```diff
@@ -0,0 +1,92 @@
+
+# Changes the tool file used by the controller for (Moto)ROS-controlled motions.
+#
+# NOTE 1: this changes the active tool for (Moto)ROS-controlled motions ONLY.
+#
+#         Regular INFORM programs, jogging or any other (programmed) motions
+#         executed by the controller will NOT use the new tool.
+#
+# NOTE 2: any tool switch initiated by invoking this service WILL however be
+#         taken into account by safety systems such as the FSU and PFL while
+#         executing (Moto)ROS-initiated motions.
+#
+# NOTE 3: active tool will be switched AFTER the controller has completed
+#         execution of the last trajectory segment that was sent to the
+#         controller BEFORE this service was called.
+#
+#         The motion queue internal to MotoROS caches a number of segments in
+#         a FIFO. Only segments received AFTER this service was invoked will
+#         be executed with the new tool. Any segments received before a tool
+#         switch was invoked will use the old tool.
+#
+#         To guarantee motion will be executed using the new tool, applications
+#         should check the 'in_motion' field (part of the RobotStatus message
+#         published on the 'robot_status' topic by the driver) and invoke the
+#         service when the robot has come to a stop (and the motion queue of
+#         MotoROS is empty). Any subsequent trajectories will use the new tool.
```

Comment:
I don't see any other way to "guarantee" a tool file change is taken into account for any particular motion with the current definition and implementation of the MotoROS service.

Reply:
I agree

Reply:
We could consider publishing a MotoROS specific message (on a ROS topic) which includes information about the MotoROS-internal queue. That could allow clients to call `SELECT_TOOL` earlier (as there would be no need for checking `in_motion`).

However MotoROS currently only returns information about the queue via a SimpleMessage service, which would perhaps be a bit too much overhead for periodic use.